File: baekjoon_all/16000+/boj_16134.cpp
// Solve 2023-03-08
// Update 2025-03-09

#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
#define SETW(n, c) cout << setw(n) << setfill(c);
#define SETP(n) cout << fixed << setprecision(n);

using namespace std;
using ll = long long;
using uint = unsigned int;
using ull = unsigned long long;
using ld = long double;
using pii = pair<int, int>;

const ll MOD = 1000000007;

ll get_factorial(ll k) {
    ll res = 1;

    for (ll i = 2; i <= k; i++) {
        res = res * i % MOD;
    }

    return res;
}

ll get_power(ll a, int b) {
    ll res = 1;

    while (b > 0) {
        if (b & 1) {
            res = res * a % MOD;
        }

        a = a * a % MOD;
        b >>= 1;
    }

    return res;
}

int main() {
    FASTIO;

    ll n, r;
    cin >> n >> r;

    ll ans = get_power(get_factorial(n - r) * get_factorial(r) % MOD, MOD - 2) * get_factorial(n) % MOD;
    cout << ans << '\n';

    return 0;
}

File: baekjoon_all/16000+/boj_16134.java
// Solve 2024-04-01
// Update 2025-03-09

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_16134 {

    static final int MOD = 1_000_000_007;

    static long[] createFactorialArr(int n) {
        long[] factorialArr = new long[n + 1];
        factorialArr[0] = factorialArr[1] = 1;

        for (int i = 2; i <= n; i++) {
            factorialArr[i] = factorialArr[i - 1] * i % MOD;
        }

        return factorialArr;
    }

    static long getPower(long a, int b) {
        long res = 1;

        while (b > 0) {
            if ((b & 1) == 1) {
                res = res * a % MOD;
            }

            a = a * a % MOD;
            b >>= 1;
        }

        return res;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());
        long[] factorialArr = createFactorialArr(n);

        long ans = getPower(factorialArr[n - r] * factorialArr[r] % MOD, MOD - 2) * factorialArr[n] % MOD;
        System.out.println(ans);
    }

}
